boot/bootstate20.go

```diff
@@ -21,329 +21,511 @@ package boot
+	}
+
+	ks20.ebl = ebl
+
+	// also get the kernel_status
+	m, err := ebl.GetBootVars("kernel_status")
+	if err != nil {
+		return err
+	}
+
+	ks20.kernelStatus = m["kernel_status"]
+
+	return nil
+}
+
+func (ks20 *bootState20Kernel) revisions() (curSnap, trySnap snap.PlaceInfo, tryingStatus string, err error) {
+	var bootSn, tryBootSn snap.PlaceInfo
+	err = ks20.loadBootenv()
+	if err != nil {
+		return nil, nil, "", err
 	}
 
 	// get the kernel for this bootloader
-	bootSnap, err = ebl.Kernel()
+	bootSn, err = ks20.ebl.Kernel()
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return nil, nil, "", fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
-	tryKernel, tryKernelExists, err := ebl.TryKernel()
+	tryKernel, tryKernelExists, err := ks20.ebl.TryKernel()
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return nil, nil, "", fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
 	if tryKernelExists {
-		tryBootSnap = tryKernel
+		tryBootSn = tryKernel
 	}
 
-	m, err := bl.GetBootVars("kernel_status")
+	return bootSn, tryBootSn, ks20.kernelStatus, nil
+}
+
+func (ks20 *bootState20Kernel) setNext(next snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
+	nextStatus, err := genericSetNext(ks20, next)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+		return false, nil, err
 	}
+	// if we are setting a snap as a try snap, then we need to reboot
+	rebootRequired = false
+	if nextStatus == TryStatus {
+		ks20.tryKernelSnap = next
+		rebootRequired = true
+	}
+	ks20.commitKernelStatus = nextStatus
 
-	trying = m["kernel_status"] == "trying"
-
-	return bootSnap, tryBootSnap, trying, nil
+	return rebootRequired, ks20, nil
 }
 
-func (s20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
-	var u20 *bootStateUpdate20
-	var err error
-	u20, err = newBootStateUpdate20(update, snap.TypeKernel)
+func (ks20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	// call the generic method with to do most of the legwork
+	u, sn, shouldClean, err := genericMarkSuccessful(ks20, update)
 	if err != nil {
 		return nil, err
 	}
 
-	// kernel_status goes from "" -> "try" -> "trying" (set by the boot script) -> ""
-	// so if we are not in "trying" mode, nothing to do here
-	if u20.env["kernel_status"] != "trying" {
-		return u20, nil
+	// if we were told to clean up status, then do that
+	if shouldClean {
+		logger.Noticef("kernel_status is %q, but there is no try kernel snap on bootloader %s, resetting to %q", ks20.kernelStatus, ks20.ebl.Name(), DefaultStatus)
+		ks20.commitKernelStatus = DefaultStatus
```

Comment:
`u.commitKernelStatus` ?

Reply:
I ended up dropping this whole part around `shouldClean` and instead unconditionally reset base_status and kernel_status to default in the commit() method on bootState20MarkSuccessful. This is simpler and requires less book-keeping like this.